File: tools/test-proxy/startup-scripts/start-server.py
import subprocess
import argparse
import json
import sys
import os

CONTAINER_NAME = "ambitious_azsdk_test_proxy"
# this image should be pinned when merged to main in your repo.
IMAGE_SOURCE = "azsdkengsys.azurecr.io/engsys/testproxy-lin:1035186"

def get_proxy_container():
    unparsed_result = subprocess.check_output(["docker", "container", "ls", "-a", "--format", "\"{{ json . }}\"", "--filter", "name={}".format(CONTAINER_NAME)])

    if unparsed_result:
        decoded = unparsed_result.decode('UTF-8').strip().replace("\\\"", "")[1:-1]
        obj_result = json.loads(decoded)
    else:
        obj_result = None

    return obj_result

def docker_interact(mode, root):
    try:
        subprocess.check_output(["docker", "--version"])
    except:
        print("An invocation of docker --version failed. This indicates that docker is not properly installed or running.")
        print("Please check your docker invocation and try running the script again.")
        sys.exit(1)

    resolved_root = os.path.abspath(root)

    if mode == "start":
        proxy_container = get_proxy_container()

        # if we already have one, we just need to check the state
        if proxy_container:
            if proxy_container["State"] == "running":
                print("Discovered an already running instance of the test-proxy!. Exiting")
                sys.exit(0)
        
        # else we need to create it
        else:
            print("Attempting creation of Docker host {}".format(CONTAINER_NAME))
            subprocess.check_call(["docker", "container", "create", "-v", "{}:/srv/testproxy/".format(resolved_root), "-p", "5001:5001", "-p", "5000:5000", "--name", CONTAINER_NAME, IMAGE_SOURCE ])

        print("Attempting start of Docker host {}".format(CONTAINER_NAME))
        subprocess.check_call(["docker", "container", "start", CONTAINER_NAME])
    
    if mode == "stop":
        proxy_container = get_proxy_container()

        if proxy_container:
            if proxy_container["State"] == "running":
                print("Found a running instance of $CONTAINER_NAME, shutting it down.")
                subprocess.check_call(["docker", "container", "stop", CONTAINER_NAME])

if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Used to define simple interface with docker CLI to handle start/stop of the test-proxy server."
    )
    
    parser.add_argument(
        "-m",
        "--mode",
        dest="mode",
        help="",
        default=".",
    )

    parser.add_argument(
        "-t",
        "--target",
        dest="target_folder",
        help="",
        default=".",
    )
   
    args = parser.parse_args()
    docker_interact(args.mode, args.target_folder)

File: tools/test-proxy/scripts/startup-scripts/start-server.js
const { exec, execSync } = require('child_process');
const { exit } = require('process');
const path = require('path');

var CONTAINER_NAME = "ambitious_azsdk_test_proxy"
// this image should be pinned when merged to main in your repo.
var IMAGE_SOURCE = "azsdkengsys.azurecr.io/engsys/testproxy-lin:1035186"

function getProxyContainer(){
    var result = execSync('docker container ls -a --format "{{ json . }}" --filter name=' + CONTAINER_NAME, (error, stdout, stderr) => {
        if (error) {
          console.error('Unable to execute docker container ls invocation.');
          console.error(error.message);
          exit(1);
        }
      
        if (stderr) {
          console.error('Something went wrong during docker container ls invocation.');
          console.error(stderr);
          exit(1);
        }
        
        return stdout;
    });

    if (result.toString()) {
        return JSON.parse(result.toString());
    }
    else {
        return undefined;
    }
}

function dockerInteract(mode, root){
    execSync('docker --version', (error, stdout, stderr) => {
        if (error || stderr) {
            console.error('A invocation of docker --version failed. This indicates that docker is not properly installed or running.');
            console.error('Please check your docker invocation and try running the script again.');

            exit(1);
        }
    });

    var resolvedPath = path.resolve(root)
    var proxyContainer = getProxyContainer();

    if(mode == "start"){
        // # if we already have one, we just need to check the state
        if(proxyContainer){
            if(proxyContainer.State == 'running'){
                console.log('Discovered an already running instance of the test-proxy!. Exiting.');
                exit(0);
            }
        }
        // else we need to create it
        else {
            console.log("Attempting creation of Docker host " + CONTAINER_NAME);
            console.log('docker container create -v ' + resolvedPath + ':/srv/testproxy/ -p 5001:5001 -p 5000:5000 --name ' + CONTAINER_NAME + ' ' + IMAGE_SOURCE);
            execSync('docker container create -v ' + resolvedPath + ':/srv/testproxy/ -p 5001:5001 -p 5000:5000 --name ' + CONTAINER_NAME + ' ' + IMAGE_SOURCE, (error, stdout, stderr) => {
                if (error) {
                    console.error('Creation of the test-proxy failed with docker error.');
                    console.error(error)
        
                    exit(1);
                }
            });
        }
        
        console.log('Attempting start of Docker host ' + CONTAINER_NAME)
        execSync('docker container start ' + CONTAINER_NAME, (error, stdout, stderr) => {
            if (error) {
                console.error('Start of the test-proxy failed with docker error.');
                console.error(error)
    
                exit(1);
            }
        });
    }

    if(mode == 'stop'){
        if (proxyContainer){
            if(proxyContainer.State == 'running'){
                console.log('Found a running instance of ' + CONTAINER_NAME + ', shutting it down.')

                execSync('docker container stop ' + CONTAINER_NAME, (error, stdout, stderr) => {
                    if (error) {
                        console.error("Stop of the test-proxy failed with docker error.");
                        console.error(error)
            
                        exit(1);
                    }
                });
            }
        }
    }
}



let args = process.argv.slice(2);

// simple error checking

if(args.length != 2){
    console.error("Unable to invoke. Expecting two arguments in order of [node start-server.js <MODE> <ROOT>]")
}

dockerInteract(args[0], args[1]);
